### PR TITLE
Wallet Connect v2

### PR DESCRIPTION
**Pull request by estebanmino**
Implementation of Wallet Connect v2

We need to wait for Layer 2 work to be merged to merge this. I'll switch the base branch to `develop` when that happens.

### CODE REVIEW

File: src/components/discover-sheet/DiscoverHome.js [hunk 1]
```diff
@@ -13,6 +14,7 @@ export default function DiscoverHome() {
     <React.Fragment>
       <TopMoversSection />
       <PulseIndex />
+      <ConnectedDapps />
```
Comment (estebanmino) [inline src/components/discover-sheet/DiscoverHome.js:17]: remember to remove this

**Comment by estebanmino** [comment]
Ready for review

WC v2 https://react-app.walletconnect.org/
WC v1 https://example.walletconnect.org/

🚀

**Comment by estebanmino** [comment]
Fixes RNBW-1434

File: src/App.js [hunk 2]
```diff
@@ -90,6 +91,10 @@ class App extends Component {
     }
     this.identifyFlow();
     AppState.addEventListener('change', this.handleAppStateChange);
+
+    // Wallet connect v2
+    walletConnectInit(store);
```
Comment (brunobar79) [inline src/App.js:96]: I guess that instead of passing it as a param you could import it directly on that file

File: src/hooks/useScanner.js [hunk 3]
```diff
@@ -136,13 +137,21 @@ export default function useScanner(enabled) {
...
-        await walletConnectOnSessionRequest(qrCodeData, () => {
-          setTimeout(enableScanning, 2000);
-        });
-      } catch (e) {
-        logger.log('walletConnectOnSessionRequest exception', e);
-        setTimeout(enableScanning, 2000);
+      const { query, pathname } = new URL(qrCodeData);
+      if (qrCodeData && query) {
+        const [, version] = pathname.split('@');
+        if (version === '2') {
```
Comment (brunobar79) [inline src/hooks/useScanner.js:143]: this "2" is in a few places, should we move it to an enum WC_VERSION_NUMBER instead?

File: src/model/walletConnect.ts [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+    version: WC_VERSION_2,
+    ...opts,
+  });
+
+const SUPPORTED_MAIN_CHAINS = [
+  'eip155:1',
+  'eip155:10',
+  'eip155:137',
+  'eip155:42161',
+];
```
Comment (brunobar79) [inline src/model/walletConnect.ts:47]: I think it's time to add the chainId inside `networkInfo.js` and then you can filter out all the networks with testnet === true

File: src/model/walletConnect.ts [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+};
+
+let client: WalletConnectClient;
+
+export const walletConnectInit = async (store: any) => {
+  if (!client) {
+    client = await WalletConnectClient.init({
+      controller: true,
+      metadata: RAINBOW_METADATA,
+      relayProvider: 'wss://relay.walletconnect.org',
```
Comment (brunobar79) [inline src/model/walletConnect.ts:91]: Might wanna move this as a constant somewhere else

File: src/model/walletConnect.ts [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+                      topic,
+                    };
+                    await client.respond(response);
+                  }
+                  return client.session.values;
+                },
+                chainId,
+                meta: metadata,
+                type: walletConnectApprovalSheetTypes.switch_chain,
+                version: 'v2',
```
Comment (brunobar79) [inline src/model/walletConnect.ts:238]: WC_VERSION_2 ?

File: src/model/walletConnect.ts [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+export const walletConnectPair = async (uri: string) => {
+  while (!client) {
+    delay(300);
+  }
+  await client.pair({ uri });
+  Navigation.handleAction(Routes.WALLET_CONNECT_APPROVAL_SHEET, {
+    callback: () => null,
+    chainId: null,
+    meta: null,
+    version: 'v2',
```
Comment (brunobar79) [inline src/model/walletConnect.ts:402]: WC_VERSION_2

File: src/screens/TransactionConfirmationScreen.js [hunk 8]
```diff
@@ -385,62 +392,62 @@ export default function TransactionConfirmationScreen() {
...
-        'Please go back to the dapp and reconnect it to your wallet',
-        [
-          {
-            onPress: () => onCancel(),
-          },
-        ]
-      );
-    }
-  }, [goBack, onCancel, peerId, walletConnector]);
+  // useEffect(() => {
```
Comment (brunobar79) [inline src/screens/TransactionConfirmationScreen.js:438]: do we wanna keep this or delete it?

**Review by brunobar79 — APPROVED** [review]
🤯 🚀 🌔

File: src/model/walletConnect.ts [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+              }
+              return client.session.values;
+            },
+            chainId: fromEIP55Format(chains?.[0]),
+            meta: {
+              dappName: metadata.name,
+              dappUrl: metadata?.url,
+              imageUrl: metadata.icons?.[0],
+            },
+            version: 'v2',
```
Comment (brunobar79) [inline src/model/walletConnect.ts:169]: WC_VERSION_2

File: src/model/walletConnect.ts [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+          });
+        } catch (error) {
+          logger.log('Exception during wc session.proposal');
+          analytics.track('Exception on wc session.proposal', {
+            error,
+            version: 'v2',
```
Comment (brunobar79) [inline src/model/walletConnect.ts:175]: WC_VERSION_2

File: src/model/walletConnect.ts [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+                    const response = {
+                      response: {
+                        id: request.id,
+                        jsonrpc: '2.0',
+                      },
+                      topic,
+                    };
+                    await client.respond(response);
+                  } else {
+                    // wcTrack('Rejected new WalletConnect session', metadata);
```
Comment (brunobar79) [inline src/model/walletConnect.ts:225]: remove?

File: src/model/walletConnect.ts [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+                  transactionDetails: requestToApprove,
+                });
+              });
+            }
+          }
+        } catch (error) {
+          logger.log('Exception during wc session.request');
+          analytics.track('Exception on wc session.request', {
+            error,
+            version: 'v2',
```
Comment (brunobar79) [inline src/model/walletConnect.ts:320]: WC_VERSION_2

**Review by brunobar79 — APPROVED** [review]

File: src/model/walletConnect.ts (hunk 6 above)
Comment (estebanmino) [reply src/model/walletConnect.ts:238]: oh so weird, changing it

File: src/model/walletConnect.ts (hunk 12 above)
Comment (estebanmino) [reply src/model/walletConnect.ts:320]: 😭

**Pull request closed without merging**